Improve multiple strings in en-US locale

n0toose opened the pull request:
I kept sending pull requests that consisted of one-line changes. It's time to
settle this once and for all. (Maybe.)

- Explain Gitea behavior and the consequences of each
  setting better, so that the user does not have to consult
  the docs.
- Do not use different spellings of identical terms
  interchangeably, e.g. `e-mail` and `email`.
- Use more conventional terms to describe the same things,
  e.g. `Confirm Password` instead of `Re-Type Password`.
- Introduces additional clarification for Mirror Settings
- Small adjustments in test
- This is a cry for help.
- Grammar and spelling consistencies for en-US locale
  (e.g. cancelled -> canceled)
- Introduce tooltip improvements.

n0toose commented:
I understand that all of these changes at once make for a stressful review process, but I figured that incremental fixes (as I have done so far) probably take up more resources on both ends here. I tried to go for being consistent; I think my biggest hesitation here involved the way I appended the word "successfully".

I'm a details person; I hope that this will make Gitea look more polished. I will not be available in the next few days (as in, it might take at least 12 hours for me to see replies), so feel free to merge any suggestions without my permission.

wxiaoguang commented on options/locale/locale_en-US.ini:
```diff
@@ -609,7 +609,7 @@ form.name_chars_not_allowed = User name "%s" contains invalid characters.
 profile = Profile
 account = Account
-appearance = Appearance
+appearance = Interface
```
I think `Appearance` is better and it is widely used.

wxiaoguang commented on options/locale/locale_en-US.ini:
```diff
@@ -360,7 +360,7 @@ remember_me = Remember This Device
 forgot_password = Forgot password?
 sign_up_now = Need an account? Register now.
-sign_up_successful = Account was successfully created.
+sign_up_successful = Account successfully created. Welcome!
```
IIRC, the word "create" is used as: "create something" or "something is created"

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -170,7 +170,7 @@ string.desc = Z - A
 [error]
 occurred = An error occurred
-report_message = If you are sure this is a Gitea bug, please search for issues on <a href="https://github.com/go-gitea/gitea/issues" target="_blank">GitHub</a> or open a new issue if necessary.
+report_message = If believe that this is a Gitea bug, please search for issues on <a href="https://github.com/go-gitea/gitea/issues" target="_blank">GitHub</a> or open a new issue if necessary.
```
```suggestion
report_message = If you believe that this is a Gitea bug, please search for issues on <a href="https://github.com/go-gitea/gitea/issues" target="_blank">GitHub</a> or open a new issue if necessary.
```

---

Oh, now is also a good time to get rid of the hardcoded link:

```suggestion
report_message = If you believe that this is a Gitea bug, please search for issues on <a href="%[1]s" target="_blank">GitHub</a> or open a new issue if necessary.
```

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -622,11 +622,11 @@ delete = Delete Account
 
 public_profile = Public Profile
-biography_placeholder = Tell us a little bit about yourself
+biography_placeholder = Tell us a little bit about yourself! (Markdown enabled)
```
```suggestion
biography_placeholder = Tell us a little bit about yourself! (You can use Markdown)
```

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -849,18 +849,18 @@ oauth2_application_create_description = OAuth2 applications gives your third-par
 
 authorized_oauth2_applications = Authorized OAuth2 Applications
-authorized_oauth2_applications_description = You've granted access to your personal Gitea account to these third party applications. Please revoke access for applications no longer needed.
+authorized_oauth2_applications_description = You have granted access to your personal Gitea account to these third party applications. Please revoke access for applications no longer needed.
```
```suggestion
authorized_oauth2_applications_description = You have granted access to your personal Gitea account to these third party applications. Please revoke access for applications you no longer need.
```

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -849,18 +849,18 @@ oauth2_application_create_description = OAuth2 applications gives your third-par
 twofa_disable = Disable Two-Factor Authentication
 twofa_scratch_token_regenerate = Regenerate Scratch Token
-twofa_scratch_token_regenerated = Your scratch token is now %s. Store it in a safe place.
+twofa_scratch_token_regenerated = Your scratch token is now %s. Please store it in a safe place now, as it will not be shown again later.
```
```suggestion
twofa_scratch_token_regenerated = Your scratch token is now %s. Store it in a safe place, it will never be shown again.
```

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -1027,8 +1027,8 @@ template.issue_labels = Issue Labels
 
-archive.title = This repo is archived. You can view files and clone it, but cannot push or open issues/pull-requests.
-archive.title_date = This repository has been archived on %s. You can view files and clone it, but cannot push or open issues/pull-requests.
+archive.title = This repo is archived. You can view files and clone it, but cannot push or open issues and pull requests.
+archive.title_date = This repository has been archived on %s. You can view files and clone it, but cannot push or open issues and pull requests.
```
```suggestion
archive.title = This repo is archived. You can view files and clone it, but cannot push or open issues or pull requests.
archive.title_date = This repository has been archived on %s. You can view files and clone it, but cannot push or open issues or pull requests.
```

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -1564,7 +1564,7 @@ issues.due_date_form = "yyyy-mm-dd"
 issues.due_date_form_edit = "Edit"
 issues.due_date_form_remove = "Remove"
-issues.due_date_not_writer = "You need repository write access to update an issue's due date."
+issues.due_date_not_writer = "You need write access to this repository in order to update an issue's due date."
```
```suggestion
issues.due_date_not_writer = "You need write access to this repository in order to update the due date of an issue."
```

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -1723,7 +1723,7 @@ pulls.no_merge_desc = This pull request cannot be merged because all repository
 pulls.no_merge_wip = This pull request cannot be merged because it is marked as being a work in progress.
 pulls.no_merge_not_ready = This pull request is not ready to be merged, check review status and status checks.
-pulls.no_merge_access = You are not authorized to merge this pull request.
+pulls.no_merge_access = You are not authorized to merge This pull request.
```
Did you use a regex?
```suggestion
pulls.no_merge_access = You are not authorized to merge this pull request.
```

n0toose replied:
Presumably unsuccessfully, I tried to go over it once again after I noticed that I messed some things up but I did not correct everything back, it seems.

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -2308,17 +2308,17 @@ settings.matrix.room_id = Room ID
 settings.archive.button = Archive Repo
 settings.archive.header = Archive This Repo
-settings.archive.text = Archiving the repo will make it entirely read-only. It is hidden from the dashboard, cannot be committed to and no issues or pull-requests can be created.
+settings.archive.text = Archiving the repo will make it entirely read-only. It will be hidden from the dashboard. Nobody (not even you!) will be able to make new commits, issues and pull requests.
```
```suggestion
settings.archive.text = Archiving the repo will make it entirely read-only. It will be hidden from the dashboard. Nobody (not even you!) will be able to make new commits, or open any issues or pull requests.
```

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -2308,17 +2308,17 @@ settings.matrix.room_id = Room ID
-settings.unarchive.text = Un-Archiving the repo will restore its ability to receive commits and pushes, as well as new issues and pull-requests.
-settings.unarchive.success = The repo was successfully un-archived.
-settings.unarchive.error = An error occurred while trying to un-archive the repo. See the log for more details.
+settings.unarchive.button = Unarchive Repo
+settings.unarchive.header = Unarchive This Repo
```
Why titlecase all of a sudden?
```suggestion
settings.unarchive.button = Unarchive repo
settings.unarchive.header = Unarchive this repo
```

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -2442,7 +2442,7 @@ release.edit_release = Update Release
 release.delete_tag = Delete Tag
 release.deletion = Delete Release
-release.deletion_desc = Deleting a release only removes it from Gitea. Git tag, repository contents and history remain unchanged. Continue?
+release.deletion_desc = Deleting a release only removes it from Gitea. The Git tag, your repository's contents and its history will not be affected. Continue?
```
```suggestion
release.deletion_desc = Deleting a release only removes it from Gitea. It will not affect the Git tag, the contents of your repository or its history. Continue?
```

n0toose commented:
Thanks for the additional recommendations and sorry for that regex fail, really goes on to show how this is super subjective work.

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -170,7 +170,7 @@ string.desc = Z - A
 [error]
 occurred = An error occurred
-report_message = If believe that this is a Gitea bug, please search for issues on <a href="https://github.com/go-gitea/gitea/issues" target="_blank">GitHub</a> or open a new issue if necessary.
+report_message = If you believe that this is a Gitea bug, please search for issues on <a href="%[1]s" target="_blank">GitHub</a> or open a new issue if necessary.
```
Of course, if we extract out the link, we need to fix it as well in
```go
templates/status/500.tmpl
45:                                     {{if .SignedUser.IsAdmin}}<p>{{.locale.Tr "error.report_message" | Str2html}}</p>{{end}}
```

n0toose replied:
I am not sure how I can approach this best.

delvh replied:
I'd guess by replacing the line with
```tmpl
                                     {{if .SignedUser.IsAdmin}}<p>{{.locale.Tr "error.report_message" "https://github.com/go-gitea/gitea/issues" | Str2html}}</p>{{end}}
```

n0toose replied:
Sorry, overcomplicated things and started thinking about whether this should be a config option or whatever. Yeah, makes sense.

wxiaoguang commented on options/locale/locale_en-US.ini:
```diff
@@ -170,7 +170,7 @@ string.desc = Z - A
 [error]
 occurred = An error occurred
-report_message = If you are sure this is a Gitea bug, please search for issues on <a href="https://github.com/go-gitea/gitea/issues" target="_blank">GitHub</a> or open a new issue if necessary.
+report_message = If you believe that this is a Gitea bug, please search for issues on <a href="%[1]s" target="_blank">GitHub</a> or open a new issue if necessary.
```
It seems strange that the URL comes from template string but the link text is hard-coded here.

The GitHub should always mean "https://github.com/go-gitea/gitea/issues", right?

I would suggest don't change it at the moment.

n0toose replied:
I also made GitHub a parameter, should I also make `Gitea` a parameter while I'm at it? I presume that you do not provide support for soft-forks like Forgejo, so it could make sense.

Of course, I'm willing to undo that change completely if possible and discuss this in a separate PR - Causing tension is not my intention (oh hey, that rhymes!).

wxiaoguang replied:
Never mind, either is fine to me as long as it makes sense. Thank you.

wxiaoguang commented on options/locale/locale_en-US.ini:
```diff
@@ -691,7 +691,7 @@ emails = Email Addresses
 manage_themes = Select default theme
 manage_openid = Manage OpenID Addresses
-email_desc = Your primary email address will be used for notifications and other operations.
+email_desc = Your primary email address will be used for notifications and Git operations made using the web interface.
```
The sentence reads strange, because IMO:

1. Not only "Git operations made using the web interface", but all related "Git operations"
2. Not only notifications, but I think there are also emails for verification or password recovery?

n0toose replied:
> Not only "Git operations made using the web interface", but all related "Git operations"

If I make a commit using a secondary email that is not primary, my primary email address will not be used.

> Not only notifications, but I think there are also emails for verification or password recovery?

Good point.

wxiaoguang replied:
> > Not only "Git operations made using the web interface", but all related "Git operations"
> 
> If I make a commit using a secondary email that is not primary, my primary email address will not be used.

I see your point and I misunderstood it before. I think it means "this email will be used as the committer email for web Git operations"

n0toose replied:
I'll change it to web-based, I think it's very much clear as day and I don't think it will confuse more advanced users that would presumably use the `tea` CLI tool later on.

n0toose replied:
32ce04ce28078464d5248162e3ab479d1ba23394 makes the text much longer, but I think that the consequences of email-related settings are very well defined. I hope that we won't have to revisit this later down the line.

wxiaoguang replied:
The sentence `will be used for .... and, provided that it is not hidden, web-based Git operations` reads a little difficult to me. But I am not a native English speaker, so I can't comment on it (aka, no more option from my side).

n0toose replied:
I don't think I can come up with a better combination of being accurate, easy to read and being transparent.

wxiaoguang commented on options/locale/locale_en-US.ini:
```diff
@@ -849,18 +849,18 @@ oauth2_application_create_description = OAuth2 applications gives your third-par
 revoke_oauth2_grant = Revoke Access
 revoke_oauth2_grant_description = Revoking access for this third party application will prevent this application from accessing your data. Are you sure?
-revoke_oauth2_grant_success = You've revoked access successfully.
+revoke_oauth2_grant_success = Access revoked successfully.
```
IIRC "revoke" is a "vt": revoke something or something is revoked

n0toose replied:
I'm trying make this string consistent with other messages like `pulls.merged_success`, `pulls.required_status_check_failed`, `pulls.update_branch_success`, `settings.transfer.success`, `settings.archive.success`.

These examples are a bit more detailed, but I found it better than changing the string to something like "You have successfully revoked access to this application." or "This application's access was revoked".

I also thing that `you've` should be removed, I'll make another commit for that.

wxiaoguang commented on options/locale/locale_en-US.ini:
```diff
@@ -1772,7 +1772,7 @@ pulls.auto_merge_has_pending_schedule = %[1]s scheduled this pull request to aut
 pulls.auto_merge_cancel_schedule = Cancel auto merge
 pulls.auto_merge_not_scheduled = This pull request is not scheduled to auto merge.
-pulls.auto_merge_canceled_schedule = The auto merge was canceled for this pull request.
+pulls.auto_merge_canceled_schedule = The auto merge was cancelled for this pull request.
```
Actually, either is fine:

https://www.grammarly.com/blog/canceled-vs-cancelled/

> Canceled or cancelled is the past tense of the verb to cancel. Both spellings are correct; Americans favor canceled (one l), while cancelled (two l's) is preferred in British English and other dialects. Canceling/cancelling and canceler/canceller follow the same pattern.

n0toose replied:
This is `en-US`, I'll replace `cancelled` in other places as well.

n0toose commented on options/locale/locale_en-US.ini:
```diff
@@ -2463,7 +2463,7 @@ branch.already_exists = A branch named "%s" already exists.
 branch.delete = Delete Branch "%s"
 branch.delete_html = Delete Branch
-branch.delete_desc = Deleting a branch is permanent. Although the deleted branch may exist for a short time before cleaning up, in most cases it CANNOT be undone. Continue?
+branch.delete_desc = Deleting a branch is permanent. Although the deleted branch may continue to exist for a short time before it actually gets removed, it CANNOT be undone in most cases. Continue?
```
That's a bad grammar mistake, I'll commit that.

wxiaoguang commented on options/locale/locale_en-US.ini:
```diff
@@ -403,7 +403,7 @@ openid_connect_title = Connect to an existing account
 openid_register_title = Create new account
 openid_register_desc = The chosen OpenID URI is unknown. Associate it with a new account here.
-openid_signin_desc = Enter your OpenID URI. For example: https://anne.me, bob.openid.org.cn or gnusocial.net/carry.
+openid_signin_desc = Enter your OpenID URI. For example: https://bob.openid.org.cn or gnusocial.net/carry.
```
I think we can use Wikipedia's example if we'd like to improve it:  https://en.wikipedia.org/wiki/OpenID

```suggestion
openid_signin_desc = Enter your OpenID URI. For example: alice.openid.example.org or https://openid.example.org/alice.
```

n0toose replied:
Good idea. I think I'm done here for the rest of this decade (or until I hear about more users "doing things wrong").

denyskon reviewed (approved):
Thank you for these changes!

denyskon commented:
@n0toose Please fix the test

n0toose commented:
> Please fix the test

Sorry, I mistakenly assumed that my changes were unrelated to the test fail. This should be fixed now (hopefully).

n0toose commented:
### Additional Screenshots

![image](https://github.com/go-gitea/gitea/assets/30193966/02273ab7-dc72-4c7a-ac4b-1454462b51b4)

![image](https://github.com/go-gitea/gitea/assets/30193966/f5b1f10b-e793-4f00-a37d-391c10c98bdc)

The second change was made as a follow-up to a previous PR of mine, https://github.com/go-gitea/gitea/pull/24433, after a user reported that the lack of `settings.mirror_settings.docs.more_information_if_disabled` was confusing, as immediately referring to the title "How do I mirror repositories?" without any given context gave the false impression that you could, indeed, still mirror repositories, even after the sentence saying that the administrator has disabled this functionality. This new string is only shown when either a pull or a push mirror is disabled as some sort of a bridge in order to signify to the user that the information contained in the following document is more generalized: This may not be obvious to users in public instances. (So, I did not have to change the substance, as in, the title of the document.)

n0toose commented:
I don't think I can get myself to stop expanding the scope of this PR.

![image](https://github.com/go-gitea/gitea/assets/30193966/3a6ba6a2-b6cc-4d12-a5b7-46c845c806e0)

wxiaoguang commented:
This review hasn't been addressed yet: https://github.com/go-gitea/gitea/pull/26213/files#r1278291138

`Access revoked successfully.`

> IIRC "revoke" is a "vt": revoke something or something is revoked

If it's sure that it is correct English syntax, then it can be marked as resolved.

wxiaoguang commented on options/locale/locale_en-US.ini:
```diff
@@ -843,7 +843,7 @@ oauth2_client_id = Client ID
 oauth2_regenerate_secret = Regenerate Secret
 oauth2_regenerate_secret_hint = Lost your secret?
-oauth2_client_secret_hint = The secret won't be visible if you revisit this page. Please save your secret.
+oauth2_client_secret_hint = The secret will disappear as soon as you leave or refresh this page. Please save your secret.
```
"disappear" reads like "the secret will be unavailable / will not exist" (IIRC "disappear" means "not exist")

Maybe it could be: "The secret is shown only once on this page" ?

n0toose replied:
I also thought of that, but thought that `Please save your secret.` essentially fixes this ambiguity.

n0toose replied:
Marking this as resolved.

lunny commented on options/locale/locale_en-US.ini:
```diff
@@ -170,7 +170,7 @@ string.desc = Z - A
 [error]
 occurred = An error occurred
-report_message = If you are sure this is a Gitea bug, please search for issues on <a href="https://github.com/go-gitea/gitea/issues" target="_blank">GitHub</a> or open a new issue if necessary.
+report_message = If you believe that this is a Gitea bug, please search for issues on <a href="%[1]s" target="_blank">%[2]s</a> or open a new issue if necessary.
```
Maybe it's better to change the key to avoid other languages rendering error?

n0toose replied:
Hi, thanks for the feedback, but I'm really not sure as to how I can address this.

silverwind replied:
He meant that outdated translations which do not have the arguments may error because we pass more arguments in than the translation accepts, but I think this is something the translation function could gracefully handle.

Too much arguments should be gracefully ignored, only too few arguments is a real problem.

silverwind replied:
Actually, it may not even be an issue because as far as I understand, the change will invalidate all existing translations, so they should fallback to english until translated, and no further changes are necessary here.

silverwind commented:
Ready to merge?

n0toose commented:
From my side, I think I am done - I am not sure what to do regarding [that](https://github.com/go-gitea/gitea/pull/26213#discussion_r1278665394) though.

n0toose commented:
Alternative proposal: Revert that change as @lunny mentioned, and then make changes everywhere so that parameters will be used instead of URLs and tags soon after the release of a new version so that translators have enough time to update the strings accordingly.

silverwind commented:
I'm 95% sure it's not an issue because changing a translation in english will invalidate all existing translations and it will fall back to english until translated again.

wxiaoguang commented:
I think lunny is right (I misunderstood crowdin too)

For example, the content of `milestones.completeness` in `en_US` has changed, it gets a `<strong>` tag, but in other languages, they all still use old translation, and crowdin didn't show re-translation for this case.

lunny commented:
> I'm 95% sure it's not an issue because changing a translation in english will invalidate all existing translations and it will fall back to english until translated again.

If those are not invalid, they maybe missed to translate? I don't know whether crowdin will think that key needs to be updated for all translations. But change the key will ask other translations to change otherwise it will not work well.

silverwind commented:
This translation system sounds so broken if a change in the primary languages does not invalidate all existing translations. I think we ought to write some tooling for this that automatically deletes any changed keys in other translation files.

n0toose commented:
Asked all three of you for reviews because I either want a final confirmation that I addressed some comments correctly or because there have been a LOT of changes ever since the "Approval". I apologize in advance for the duplicated effort.

I tried my best not to address the parameter problems for the moment, as that seems like an entirely separate issue that is out of my reach as a contributor.

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -368,7 +368,7 @@ reset_password_mail_sent_prompt = A confirmation email has been sent to <b>%s</b
 account_activated = Account has been activated
 prohibit_login = Sign In Prohibited
-prohibit_login_desc = Your account is prohibited to sign in, please contact your site administrator.
+prohibit_login_desc = Your account has been banned, please contact your site administrator for more information.
```
That's not true, there are also other reasons why an account cannot sign in, i.e. it is a bot account or belongs to an organization.

n0toose replied:
ACK

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -2464,7 +2467,7 @@ branch.already_exists = A branch named "%s" already exists.
 branch.delete = Delete Branch "%s"
 branch.delete_html = Delete Branch
-branch.delete_desc = Deleting a branch is permanent. Although the deleted branch may continue to exist for a short time before it actually gets removed, it CANNOT be undone in most cases. Continue?
+branch.delete_desc = Deleting a branch is permanent. Although the deleted branch may continue existing for a short time before it actually gets purged, it CANNOT be undone in most cases. Continue?
```
I think the old one is better

n0toose replied:
Me too.

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -3369,17 +3372,17 @@ settings.delete.success = The package has been deleted.
 owner.settings.cargo.title = Cargo Registry Index
 owner.settings.cargo.initialize = Initialize Index
-owner.settings.cargo.initialize.description = To use the Cargo registry a special index git repository is needed. Here you can (re)create it with the required config.
+owner.settings.cargo.initialize.description = A special index Git repository is needed to use the Cargo registry. Using this option will (re-)create it with all required configurations.
```
Isn't `configuration` a word like `sheep` or `fish` whose plural is its singular?

n0toose replied:
This is correct. Bear in mind that I work on this PR very late at night so that I can actually handle the mundane-ness of it, which is good for progress but still results in many little errors. I'll fix this.

n0toose replied:
Hm,
- https://www.merriam-webster.com/dictionary/configuration (`configurations` example used)
- https://www.wordhippo.com/what-is/the-plural-of/configuration.html

I think we are both right, but it may be the case that using plural in the first place makes the idea that is meant to be conveyed here too inaccurate.

n0toose replied:
I tried something else here: https://github.com/go-gitea/gitea/pull/26213/commits/faf68dce4f2d5ec3caa731aedc2125ec6261207f

wxiaoguang reviewed (commented):
Thank you for inviting me to review, while I am not an English native speaker, I think people understand English better could help.

n0toose commented on options/locale/locale_en-US.ini:
```diff
@@ -1954,6 +1954,7 @@ settings.mirror_settings.docs.disabled_push_mirror.info = Push mirrors have been
 settings.mirror_settings.docs.no_new_mirrors = Your repository is mirroring changes to or from another repository. Please keep in mind that you can't create any new mirrors at this time.
 settings.mirror_settings.docs.can_still_use = Although you can't modify existing mirrors or create new ones, you may still use your existing mirror.
 settings.mirror_settings.docs.pull_mirror_instructions = To set up a pull mirror, please consult:
+settings.mirror_settings.docs.more_information_if_disabled = You can find out more about push and pull mirrors here:
```
See https://github.com/go-gitea/gitea/pull/26213/files#diff-dda0bd64abe99b2fc96238e95a74eddf6b89c1210a07130f1d6bedf7528cf461R1961.

n0toose replied:
See https://github.com/go-gitea/gitea/pull/26213/files#diff-dda0bd64abe99b2fc96238e95a74eddf6b89c1210a07130f1d6bedf7528cf461R1961.

silverwind commented:
> > I'm 95% sure it's not an issue because changing a translation in english will invalidate all existing translations and it will fall back to english until translated again.
> 
> If those are not invalid, they maybe missed to translate? I don't know whether crowdin will detec that key have been updated and ask all transations to update the key.
> 
> But changing the key can do that otherwise it will not work well.

I think it's default behaviour of crowdin to invalidate existing translation when the primary value changes. This is [this option](https://developer.crowdin.com/configuration-file/#changed-strings-update) which by default is not set, so I assume we don't set it and I recommend to never set it.

denyskon commented:
Can we merge this? We could infinitely continue adding stuff here if we wanted, but I think we should merge
